src/spinnaker.cpp

```diff
@@ -0,0 +1,1033 @@
+          const T default_key)
+{
+    for (const auto& [k, v] : table) {
+        if (v == value) {
+            return k;
+        }
+    }
+    return default_key;
+}
+
+// Maps Acquire trigger edges to GenICam activation strings.
+const std::unordered_map<TriggerEdge, Spinnaker::GenICam::gcstring>
+  trigger_edge_to_activation{
+      { TriggerEdge_Rising, "RisingEdge" },
+      { TriggerEdge_Falling, "FallingEdge" },
+      { TriggerEdge_AnyEdge, "AnyEdge" },
+      { TriggerEdge_LevelHigh, "LevelHigh" },
+      { TriggerEdge_LevelLow, "LevelLow" },
+  };
+
+TriggerEdge
+to_trigger_edge(const Spinnaker::GenICam::gcstring& activation)
+{
+    return inv_at_or(
+      trigger_edge_to_activation, activation, TriggerEdge_Unknown);
+}
+
+const Spinnaker::GenICam::gcstring&
+to_trigger_activation(TriggerEdge edge)
+{
+    return trigger_edge_to_activation.at(edge);
+}
+
+// Maps Acquire sample types to GenICam pixel format strings.
+const std::unordered_map<SampleType, Spinnaker::GenICam::gcstring>
+  sample_type_to_pixel_format{
+      { SampleType_u8, "Mono8" },    { SampleType_i8, "Mono8s" },
+      { SampleType_u10, "Mono10" },  { SampleType_u12, "Mono12" },
+      { SampleType_u14, "Mono14" },  { SampleType_u16, "Mono16" },
+      { SampleType_i16, "Mono16s" }, { SampleType_f32, "Mono32f" },
+  };
+
+SampleType
+to_sample_type(const Spinnaker::GenICam::gcstring& pixel_format)
+{
+    return inv_at_or(
+      sample_type_to_pixel_format, pixel_format, SampleType_Unknown);
+}
+
+const Spinnaker::GenICam::gcstring&
+to_pixel_format(SampleType sample_type)
+{
+    return sample_type_to_pixel_format.at(sample_type);
+}
+
+// Maps Acquire line numbers to GenICam line source strings.
+// Define the software line as acquire's line 2 because the Blackfly USB3 camera
+// only has two physical lines (0 and 1).
+const std::unordered_map<uint8_t, Spinnaker::GenICam::gcstring>
+  trigger_line_to_source{
+      { 0, "Line0" },
+      { 1, "Line1" },
+      { 2, "Software" },
+  };
+
+uint8_t
+to_trigger_line(const Spinnaker::GenICam::gcstring& source)
+{
+    // Acquire's line 3 is unassigned here so use it for an unknown line.
+    return inv_at_or<uint8_t>(trigger_line_to_source, source, 3);
+}
+
+const Spinnaker::GenICam::gcstring&
+to_trigger_source(uint8_t line)
+{
+    return trigger_line_to_source.at(line);
+}
+
+// Returns true if two triggers are equal in value, false otherwise.
+bool
+is_equal(const Trigger& lhs, const Trigger& rhs)
+{
+    return memcmp(&lhs, &rhs, sizeof(Trigger)) == 0;
+}
+
+// Clamps a value to be in the closed interval [low, high].
+template<typename T>
+T
+clamp(T value, T low, T high)
+{
+    if (value < low)
+        return low;
+    if (value > high)
+        return high;
+    return value;
+}
+
+// Returns true if a Spinnaker node is writable, false otherwise.
+bool
+check_node_writable(Spinnaker::GenApi::INode& node)
+{
+    if (!Spinnaker::GenApi::IsWritable(node)) {
```

Comment:
I've flip-flopped between guarding with an `if` and just `throwing`. I think the semantics of `set` mean that we want to try to set as many new parameter values as we can even if one fails, so I'm using the `if` guard right now. Maybe I should just wrap the individual setters in try/catch blocks and let spinnaker throw though?